### PR TITLE
support streaming audio for mod_audio_fork

### PR DESCRIPTION
**Pull request by xquanluu**
https://github.com/drachtio/drachtio-fsmrf/pull/93
https://github.com/jambonz/verb-specifications/pull/48
https://github.com/jambonz/jambonz-feature-server/pull/695

### CODE REVIEW

File: mod_audio_fork/audio_pipe.cpp [hunk 1]
```diff
@@ -178,11 +180,9 @@ int AudioPipe::lws_callback(struct lws *wsi,
             ap->m_recv_buf_ptr += len;
           }
           if (lws_is_final_fragment(wsi)) {
-            if (nullptr != ap->m_recv_buf) {
-              std::string msg((char *)ap->m_recv_buf, ap->m_recv_buf_ptr - ap->m_recv_buf);
-              ap->m_callback(ap->m_uuid.c_str(), ap->m_bugname.c_str(), AudioPipe::MESSAGE, msg.c_str());
-              if (nullptr != ap->m_recv_buf) free(ap->m_recv_buf);
-            }
+            std::string msg((char *)ap->m_recv_buf, ap->m_recv_buf_ptr - ap->m_recv_buf);
+            ap->m_callback(ap->m_uuid.c_str(), ap->m_bugname.c_str(), is_binary ? AudioPipe::BINARY : AudioPipe::MESSAGE, msg.c_str());
```
Comment (davehorton) [inline mod_audio_fork/audio_pipe.cpp:184]: this results in a copy of the binary data.  I would the expense of a copy operation here, for binary data.  (For text frames it's fine).  I would rather the insert into the circular buffer points to directly the raw incoming read buffer

**Review by davehorton — CHANGES_REQUESTED** [review]

File: mod_audio_fork/audio_pipe.cpp (hunk 1 above)
Comment (xquanluu) [reply mod_audio_fork/audio_pipe.cpp]: fixed

**Comment by xquanluu** [comment]
https://github.com/jambonz/freeswitch-modules/issues/25

File: mod_audio_fork/audio_pipe.cpp [hunk 2]
```diff
@@ -178,11 +184,9 @@ int AudioPipe::lws_callback(struct lws *wsi,
             ap->m_recv_buf_ptr += len;
           }
           if (lws_is_final_fragment(wsi)) {
```
Comment (davehorton) [inline mod_audio_fork/audio_pipe.cpp:186]: I'm not sure about this.  Why was this changed?  This has to do with text frames, not binary frames.  With text frames we can still have the same issue as before where one message is split between different frames and we need to collect and process *only* when the final frame arrives.

**Review by davehorton — CHANGES_REQUESTED** [review]

File: mod_audio_fork/mod_audio_fork.c [hunk 3]
```diff
@@ -159,10 +179,10 @@ static switch_status_t send_text(switch_core_session_t *session, char* bugname,
   return status;
 }
 
-#define FORK_API_SYNTAX "<uuid> [start | stop | send_text | pause | resume | graceful-shutdown ] [wss-url | path] [mono | mixed | stereo] [8000 | 16000 | 24000 | 32000 | 64000] [bugname] [metadata]"
+#define FORK_API_SYNTAX "<uuid> [start | stop | send_text | pause | resume | graceful-shutdown | stop_play ] [wss-url | path] [mono | mixed | stereo] [8000 | 16000 | 24000 | 32000 | 64000] [bugname] [metadata] [bidrectionalAduo_enabled] [bidrectionalAduo_stream_enabled] [bidrectionalAduo_stream_samplerate]"
```
Comment (davehorton) [inline mod_audio_fork/mod_audio_fork.c:182]: fix typos "bidrectionalAduo"

File: mod_audio_fork/audio_pipe.cpp [hunk 4]
```diff
@@ -468,7 +473,7 @@ AudioPipe::AudioPipe(const char* uuid, const char* host, unsigned int port, cons
     m_username.assign(username);
     m_password.assign(password);
   }
-
+  m_bidirectional_audio_stream = bidirectional_audio_stream > 0;
```
Comment (davehorton) [inline mod_audio_fork/audio_pipe.cpp:476]: I'm curious about this.  Why the check > 0?  why not doesn't this work?
```
m_bidirectional_audio_stream = bidirectional_audio_stream
```
the greater than check might make the reader think there is something more involved going on here.....is there?

File: mod_audio_fork/audio_pipe.cpp (hunk 2 above)
Comment (xquanluu) [reply mod_audio_fork/audio_pipe.cpp:186]: if (nullptr != ap->m_recv_buf) { 

already checked in 180

I'm reverting the code change for this

File: mod_audio_fork/audio_pipe.cpp (hunk 4 above)
Comment (xquanluu) [reply mod_audio_fork/audio_pipe.cpp:476]: fixed

File: mod_audio_fork/mod_audio_fork.c (hunk 3 above)
Comment (xquanluu) [reply mod_audio_fork/mod_audio_fork.c:182]: fixed

**Review by davehorton — APPROVED** [review]

**Pull request merged**